Major update to SKLL documentation & other fixes (#568)
EducationalTestingService/skll: doc/run_experiment.rst

    
    RANSACRegressor
        .. code-block:: python
    
           {'loss': 'squared_loss', 'random_state': 123456789}
    
    Ridge and RidgeClassifier
        .. code-block:: python
    
           {'random_state': 123456789}
    
    SVC and SVR
        .. code-block:: python
    
           {'cache_size': 1000, 'gamma': 'scale'}
    
    SGDClassifier
        .. code-block:: python
    
           {'loss': 'log', 'max_iter': 1000, random_state': 123456789, 'tol': 1e-3}

    SGDRegressor
        .. code-block:: python
    
           {'max_iter': 1000, 'random_state': 123456789, 'tol': 1e-3}
    
    TheilSenRegressor
        .. code-block:: python
    
           {'random_state': 123456789}

    .. _imbalanced_data:

    .. note::

        The `fixed_parameters` field offers us a way to deal with imbalanced
        data sets by using the parameter ``class_weight`` for the following 
        classifiers: ``DecisionTreeClassifier``, ``LogisticRegression``, 
        ``LinearSVC``, ``RandomForestClassifier``, ``RidgeClassifier``, 
        ``SGDClassifier``, and ``SVC``.

        Two possible options are available. The first one is ``balanced``,

>         which automatically adjust weights inversely proportional to class

Reviewer: "adjust" -> "adjusts"

.. _ids_to_floats:

ids_to_floats *(Optional)*
""""""""""""""""""""""""""

If you have a dataset with lots of examples, and your input files have IDs that
look like numbers (can be converted by float()), then setting this to True will
save you some memory by storing IDs as floats. Note that this will cause IDs to
be printed as floats in prediction files (e.g., ``4.0`` instead of ``4`` or
``0004`` or ``4.000``).

.. _label_col:

label_col *(Optional)*
""""""""""""""""""""""

If you're using :ref:`ARFF <arff>`, :ref:`CSV <csv>`, or :ref:`TSV <csv>`
files, the class labels for each instance are assumed to be in a column with
this name. If no column with this name is found, the data is assumed to be
unlabelled. Defaults to ``y``. For ARFF files only, this must also be the final
column to count as the label (for compatibility with Weka).

.. _learning_curve_cv_folds_list:

learning_curve_cv_folds_list *(Optional)*
""""""""""""""""""""""""""""""""""""""""""

List of integers specifying the number of folds to use for cross-validation
at each point of the learning curve (training size), one per learner. For
example, if you specify the following learners: ``["SVC", "LogisticRegression"]``,

> specifying ``[10, 100]`` as the value of ``learning_curve_cv_folds_list`` will

Reviewer: I think "it" should be added before "will". Or consider rephrasing.

tell SKLL to use 10 cross-validation folds at each point of the SVC curve and
100 cross-validation folds at each point of the logistic regression curve. Although
more folds will generally yield more reliable results, smaller number of folds
may be better for learners that are slow to train. Defaults to 10 for
each learner.

.. _learning_curve_train_sizes:

learning_curve_train_sizes *(Optional)*
""""""""""""""""""""""""""""""""""""""""""

List of floats or integers representing relative or absolute numbers
of training examples that will be used to generate the learning curve
respectively. If the type is float, it is regarded as a fraction of
the maximum size of the training set (that is determined by the selected

> validation method), i.e. it has to be within (0, 1]. Otherwise it is

Reviewer 2: ```suggestion
validation method), i.e. it has to be within [0, 1]. Otherwise it is
```

Author: I actually intended to use the left open interval since you can't really have 0 as the train size for which to plot the learning curve :)

interpreted as absolute sizes of the training sets. Note that for classification

> the number of samples usually have to be big enough to contain at least

Reviewer 2: ```suggestion
the number of samples usually has to be big enough to contain at least
```

    *   **uwk_off_by_one**: Same as ``unweighted_kappa``, but all ranking
        differences are discounted by one. In other words, a ranking of
        1 and a ranking of 2 would be considered equal.

The following is a list of the other optional fields in this section in alphabetical order.

.. _grid_search:

grid_search *(Optional)*
""""""""""""""""""""""""

Whether or not to perform grid search to find optimal parameters for

> classifier. Defaults to ``True`` since optimizing model hyperparameters

Reviewer: We should probably change "classifier" to something more general, like "learner"

almost always leads to better performance. Note that for the
:ref:`learning_curve <learning_curve>` task, grid search is not allowed
and setting it to ``True`` will generate a warning and be ignored.

.. note:: 

    1. In versions of SKLL before v2.0, this option was set to
       ``False`` by default but that was changed since the benefits

>        of hyperparameter tuning significantly outweight the cost

Reviewer: "outweight" -> "outweigh"

>        in terms of model fitting time. Instead, SKLL must explicly

Reviewer 2: ```suggestion
       in terms of model fitting time. Instead, SKLL users must explicitly
```

    
    MLPClassifier and MLPRegressor:
        .. code-block:: python
    
           [{'activation': ['logistic', 'tanh', 'relu'],
             'alpha': [1e-4, 1e-3, 1e-3, 1e-1, 1],
             'learning_rate_init': [0.001, 0.01, 0.1]}],
    
    MultinomialNB
        .. code-block:: python
    
           [{'alpha': [0.1, 0.25, 0.5, 0.75, 1.0]}]
    
    RandomForestClassifier and RandomForestRegressor
        .. code-block:: python
    
           [{'max_depth': [1, 5, 10, None]}]
    
    Ridge and RidgeClassifier
        .. code-block:: python
    
           [{'alpha': [0.01, 0.1, 1.0, 10.0, 100.0]}]
    
    SGDClassifier and SGDRegressor
        .. code-block:: python
    
            [{'alpha': [0.000001, 0.00001, 0.0001, 0.001, 0.01],
              'penalty': ['l1', 'l2', 'elasticnet']}]
    
    SVC
        .. code-block:: python
    
           [{'C': [0.01, 0.1, 1.0, 10.0, 100.0],
             'gamma': ['auto', 0.01, 0.1, 1.0, 10.0, 100.0]}]
    
    SVR
        .. code-block:: python
    
           [{'C': [0.01, 0.1, 1.0, 10.0, 100.0]}]

    .. note::
           Note that learners not listed here do not have any default

>            parameter grids in SKLL either because either there are no

Reviewer: Remove one of the "either"s

        Full names must be specified, (e.g., ``nlp.research.ets.org``).

.. _experiment_output_files:

Output files
------------

For most of the SKLL tasks the various output files generated by :ref:`run_experiment <run_experiment>` share the automatically generated prefix 
``<EXPERIMENT>_<FEATURESET>_<LEARNER>_<OBJECTIVE>``, where the following definitions hold:

    ``<EXPERIMENT>``

>         The value of the as :ref:`experiment_name` field in the configuration file.

Reviewer 2: ```suggestion
        The value of the :ref:`experiment_name` field in the configuration file.
```

    ``<FEATURESET>``
        The components of the feature set that was used for training, joined with "+".

    ``<LEARNER>``

>         The learner that was used to generate the current current results/model/etc. 

Reviewer: Remove one of the "current"s

2. Files ending in ``.results.json``, which contain all of the same information as the
   ``.results`` files, but in a format more well-suited to automated processing. In
   some cases, ``.results.json`` files may contain *more* information than their
   ``.results`` file counterparts. For example, when doing :ref:`grid search <grid_search>`
   for tuning model hyperparameters, these files contain an additional attribute ``grid_search_cv_results`` containing detailed results from the grid search process.

.. _output_prediction_files:

Prediction files
^^^^^^^^^^^^^^^^

Predictions files are TSV files that contain either the predicted
values (for regression) OR predicted labels/class probabiltiies 
(for classification) for each instance in the test feature set. 
The value of the :ref:`probability <probability>` option decides whether SKLL
outputs the labels or the probabilities.

When the predictions are labels or values, there
are only two columns in the file: one containing the ID for the instance
and the other containing the prediction. The headers for the two columns
in this case are "id" and "prediction".

When the predictions are class probabilities, there are N+1 columns

> in these files, where N are the number of classes in the training

Reviewer: "are" -> "is"
